### PR TITLE
Use custom image registry

### PR DESCRIPTION
**Pull request by xaniasd**
resolves kubeflow/kubeflow#3210

Introduces imageRegistry parameter in KfDef spec. If set, all generated kustomizations will use repositories from the specified registry. 

**NB** istio and application images are not affected

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfctl/30)

### CODE REVIEW

**Comment by jlewi** [comment]
@xaniasd thanks for taking this on. 

A couple things until #7 is resolved; development of kfctl should happen in kubeflow/kubeflow and not in kubeflow/kfctl.

Why introduce a new parameter ImageRegistry into KfDefSpec rather than introduce a new go binary  or subcommands to provide bulk manipulation of the docker images in the kustomize files?

If we introduce KfDefSpec.ImageRegistry and just rewrite the images is that going to give us the power we need?

For example, if users want to pull from a private repo or replicate their images to their own mirror then we need to find and output all the existing images so that they can be mirror'd.

**Comment by xaniasd** [comment]
hi @jlewi, thanks for your feedback.

> A couple things until #7 is resolved; development of kfctl should happen in kubeflow/kubeflow and not in kubeflow/kfctl.

thanks, noted

> Why introduce a new parameter ImageRegistry into KfDefSpec rather than introduce a new go binary or subcommands to provide bulk manipulation of the docker images in the kustomize files?

my implementation aims for replacing images, as described in the issue. Also along the lines of what the [script](https://github.com/kubeflow/kubeflow/blob/master/scripts/gke/use_gcr_for_all_images.sh) does.

> If we introduce KfDefSpec.ImageRegistry and just rewrite the images is that going to give us the power we need?
> 
> For example, if users want to pull from a private repo or replicate their images to their own mirror then we need to find and output all the existing images so that they can be mirror'd.

I probably miss some background here so I can't really answer your questions. I did find the design doc of kfctl in kubeflow/manifests#339, but a bit late I guess, please excuse my ignorance. What are the bulk manipulations we have in mind? There aren't any specified in the design doc. Furthermore, this functionality is relevant to kustomize, which according to the doc is not considered a plugin, correct?

let me know how you want to proceed with this!

**Comment by jlewi** [comment]
@xaniasd The functionality we'd like is something like the following

```
kfctl set-image-registry <old prefix> <new prefix>
```

So the behavior would be something like the following

```
cd ${KFAPP}
kfctl build ....
```
* This will create all the kustomize manifests in ${KFAPP}/kustomize

The user would then do something like

```
cd ${KFAPP}/kustomize
kfctl set-image-registry gcr.io/kubeflow-images-public  myprivate/registry
```

* This would then recourse through all the kustomization.YAML files and replace the old prefix with the new prefix.

Eventually we might provide built in support for this as part of KFDef. I don't think we want to tightly integrate into KFDef right now because we are still experimenting with the right semantics for image replacement.

For KFDef on the other hand we are trying to get the API to v1beta1 in 0.7 and then v1 in Q4. So we don't want to add new/experimental fields to it right now.

**Comment by jlewi** [comment]
@xaniasd Think you might be able to take another stab at this this week?

**Comment by xaniasd** [comment]
@jlewi I don't think I'll have time for it for the next 5-10 days, pretty busy agenda unfortunately.

**Comment by jlewi** [comment]
Thanks @xaniasd 

@gyliu513 @jinchihe Any interest in picking this up since you will need it for POWER?

**Comment by gyliu513** [comment]
@alongzhi can you help take over this?

FYI @jinchihe @jlewi

**Comment by hemantha-kumara** [comment]
> * This would then recourse through all the kustomization.YAML
@jlewi All kustomization.yaml doesn't have images.name & images.newName 
for example 
- https://github.com/kubeflow/manifests/blob/master/metadata/base/kustomization.yaml
- https://github.com/kubeflow/manifests/blob/master/istio/istio/base/kustomization.yaml  
with existing approach how 
This mean, there are two approaches to solve this problem
1. Either submit all kustmization files with have images field for all docker images
2. Parse all deployment and statefulset .yaml to find name of the image and do replace using kustomize edit set image <imagenamename>=<newName>

**Comment by alongzhi** [comment]
@gyliu513 @jlewi Okay, I'm looking at this one.

**Comment by nrchakradhar** [comment]
@alongzhi @gyliu513 @jlewi
https://github.com/kubeflow/manifests/pull/363/files
This might help.

**Comment by jlewi** [comment]
Thanks @nrchakradhar 
@xaniasd Are you planning on updating this PR per my comments above?

**Comment by xaniasd** [comment]
@jlewi sure, I'll give it a try

**Comment by xaniasd** [comment]
@jlewi @yanniszark 

here's another version in line with what we discussed, let me know what you think.

As soon as we're ok with this approach, I'll close this PR and submit another one on kubeflow/kubeflow only with the good stuff.

File: cmd/kfctl/cmd/set-image-name.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+)
+
+var filter string
+var flatten bool
+
+func init() {
+	setImageNameCmd.Flags().StringVarP(&filter, "filter", "f", "", "Only set name for images with matching prefix")
+	setImageNameCmd.Flags().BoolVar(&flatten, "flatten", false, "Set to true for registries not supporting hierarchical paths with more than two components")
```
Comment (jlewi) [inline cmd/kfctl/cmd/set-image-name.go]: Why does this mean? Can you give me example command line syntax?

**Comment by jlewi** [comment]
@xaniasd Thanks. Could you provide some examples of what the command line syntax is?

**Comment by xaniasd** [comment]
hi @jlewi, sure thing. See below:

```bash
# replace registry domain if exists, append if not
# so mysql:latest - test.io/mysql:latest
# gcr.io/kubeflow-images-public/ingress-setup:latest -> test.io/kubeflow-images-public/ingress-setup:latest 
kfctl set-image-name test.io
# same as above, but only for image names starting with gcr.io
kfctl set-image-name test.io --filter gcr.io
# gcr.io/kubeflow-images-public/ingress-setup:latest -> test.io/my-repo/kubeflow-images-public/ingress-setup:latest
kfctl set-image-name test.io/my-repo 
# gcr.io/kubeflow-images-public/ingress-setup:latest -> test.io/my-repo/ingress-setup:latest
kfctl set-image-name --flatten test.io/my-repo
```

about the effect of --flatten, you can find some more examples in the unit tests

**Comment by yanniszark** [comment]
@jlewi after discussing with @xaniasd I realized that not all image registries support subdirectories.
In fact, many support only a flat structure (eg Docker registry).
Because of that, I believe it's important to support flattening the images.

The algorithm more or less is:
1. User gives a registry (eg `test.io/kf-v062/` under which to move images.
2. For every image (eg `gcr.io/kubeflow-images/vizier/katib/katib-ui`) matching `filter`, strip the origin and:
    1. `flatten=true`: Keep only the last part of the path. (`test.io/kf-v062/katib-ui`)
    2. `flatten=false`: Keep the whole path. (`test.io/kf-v062/kubeflow-images/vizier/katib/katib-uikatib-ui`)

**Comment by jlewi** [comment]
Thanks @xaniasd @yanniszark 
semantics LGTM.

**Comment by yanniszark** [comment]
Great!
@xaniasd shall we take this to kubeflow/kubeflow?

**Comment by jlewi** [comment]
Closing this PR because the code was merged into kubeflow/kubeflow. The code has now been migrated to kubeflow/kfctl and is here
https://github.com/kubeflow/kfctl/blob/master/cmd/kfctl/cmd/set-image-name.go

**Pull request closed without merging**